lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+
+        self.units = units
+
+    def __iter__(self):
+
+        for unit in self.units:
+            yield Lock(self.rel_name, self.charm, unit=unit)
+
+
+class RollingEvent(EventBase):
+    def __init__(self, handle, name):
+        super().__init__(handle)
+        self._name = name
+
+    def snapshot(self):
+        return {"name": self._name}
+
+    def restore(self, snapshot):
+        self._name = snapshot["name"]
+
+    @property
+    def name(self):
+        """The `name` property allows us to namespace Rolling Operations."""
+        return self._name
+
+
+class RunWithLock(RollingEvent):
+    pass
+
+
+class AcquireLock(RollingEvent):
+    pass
+
+
+class ProcessLocks(RollingEvent):
+    pass
+
+
+class RollingEvents(CharmEvents):
+    run_with_lock = EventSource(RunWithLock)
+    acquire_lock = EventSource(AcquireLock)
+    process_locks = EventSource(ProcessLocks)
+
+
+class RollingOpsManager(Object):
+    """Emitters and handlers for rolling ops."""
+
+    def __init__(self, charm, rel_name, func_run, event_acquire):
+        """
+        params:
+            charm: the charm we are attaching this to.
+            rel_name: an identifier, by convention based on the name of the relation in the
+                metadata.yaml, which identifies this instance of RollingOperatorsFactory,
+                distinct from other instances that may be hanlding other events.
+            func_run: a closure to run when we have a lock. (It must take a CharmBase object and
+                EventBase object as args.)
+            event_acquire: The event that will prompt us to acquire a lock.
+        """
+        super().__init__(charm, None)
+
+        self.charm = charm
+        self.name = rel_name
+        self._func_run = func_run
+
+        self.framework.observe(event_acquire, self._on_acquire_lock)
+        self.framework.observe(charm.on[self.name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on.run_with_lock, self._on_run_with_lock)
+        self.framework.observe(charm.on.process_locks, self._on_process_locks)
+
+    def _func_run(self: CharmBase, event: EventBase) -> None:
+        """Placeholder for the function that actually runs our event.
+
+        Usually overridden in the init.
+        """
+        raise NotImplementedError
+
+    def _on_relation_changed(self: CharmBase, event: RelationChangedEvent):
+        """Process relation changed.
+
+        First, determine whether this unit has been granted a lock. If so, emit a RunWithLock
+        event.
+
+        Then, if we are the leader, fire off a process locks event.
+
+        """
+        if not event.relation.name == self.name:
+            # This is not relevant to us.
+            return
+
+        if Lock(self.name, self.charm).is_held():
+            self.charm.on.run_with_lock.emit(name=self.name)
+
+        if self.framework.model.unit.is_leader():
+            self.charm.on.process_locks.emit(name=self.name)
```

Comment:
@rwcarlsen I went back and forth on this. `self.charm.on.process_locks.emit` is definitely wordier than "self._process_locks", and the both are equally easy to override if extending this library.

But the event contains the following logic for free:

```
try:
    self._on_process_locks()
except Exception:
   self._store.run_with_lock = "already_executed"
   event.defer()
   return
else:
   self._store.run_with_lock = ""  # Clear for next time.
```

I like that a lot. Driving things as events provides a nice separation of concerns, and some clarity in the interface.

It would be a *lot* cleaner if the custom events contained their own handlers, which got automagically attached to the charm, but I'm not sure that this is the right place to establish that pattern.

Reply:
Following up on this: TIL that you can run custom events from the juju cli!

That has the potential to make all of this a lot cleaner. You can drop the rigamarole with defining and action, and just expose an event that operators (both human and software) can call on when they want to. I like this a whole lot :-)